File: electron-core/utils/updates.ts
import * as fs from "node:fs";
import * as https from "node:https";
import * as path from "node:path";

import { DISPLAY_NAME, GITHUB_ORG, GITHUB_REPO } from "./brand.js";
import { appCacheDir, ensureCacheDir } from "./paths.js";

const CACHE_DURATION_SECS = 3600;
const USER_AGENT = "metaburn-update-checker/1.0";

export interface UpdateResult {
  current_version: string;
  latest_version: string;
  update_available: boolean;
  download_url?: string | null;
  release_notes?: string | null;
  release_date?: string | null;
  error?: string | null;
}

interface GitHubRelease {
  tag_name: string;
  html_url?: string;
  body?: string;
  published_at?: string;
}

interface CachePayload {
  timestamp: number;
  latest_version: string;
  download_url?: string | null;
  release_notes?: string | null;
  release_date?: string | null;
}

function cacheFile(): string {
  return path.join(appCacheDir(), "update_check.json");
}

function nowSecs(): number {
  return Math.floor(Date.now() / 1000);
}

/** Compare major.minor.patch (optional leading `v`). Returns -1 / 0 / 1. */
export function compareVersions(a: string, b: string): number {
  const pa = parseVersion(a);
  const pb = parseVersion(b);
  if (!pa || !pb) return 0;
  for (let i = 0; i < 3; i++) {
    if (pa[i]! < pb[i]!) return -1;
    if (pa[i]! > pb[i]!) return 1;
  }
  return 0;
}

function parseVersion(version: string): [number, number, number] | null {
  const cleaned = version.trim().replace(/^v/i, "");
  const match = /^(\d+)\.(\d+)\.(\d+)/.exec(cleaned);
  if (!match) return null;
  return [Number(match[1]), Number(match[2]), Number(match[3])];
}

function readCache(): CachePayload | null {
  try {
    const raw = fs.readFileSync(cacheFile(), "utf8");
    const payload = JSON.parse(raw) as CachePayload;
    if (nowSecs() - payload.timestamp > CACHE_DURATION_SECS) return null;
    return payload;
  } catch {
    return null;
  }
}

function writeCache(payload: CachePayload): void {
  try {
    ensureCacheDir();
    fs.writeFileSync(cacheFile(), JSON.stringify(payload), "utf8");
  } catch {
    // non-fatal
  }
}

function fetchLatestRelease(): Promise<GitHubRelease> {
  const url = `https://api.github.com/repos/${GITHUB_ORG}/${GITHUB_REPO}/releases/latest`;
  return new Promise((resolve, reject) => {
    const req = https.get(
      url,
      {
        headers: {
          "User-Agent": USER_AGENT,
          Accept: "application/vnd.github.v3+json",
        },
      },
      (res) => {
        let data = "";
        res.on("data", (chunk) => {
          data += chunk;
        });
        res.on("end", () => {
          if (res.statusCode && res.statusCode >= 400) {
            reject(new Error(`GitHub Releases returned HTTP ${res.statusCode}`));
            return;
          }
          try {
            resolve(JSON.parse(data) as GitHubRelease);
          } catch (err) {
            reject(err);
          }
        });
      },
    );
    req.on("error", reject);
  });
}

/** Check GitHub Releases for a newer version of MetaBurn. */
export async function checkForUpdates(currentVersion: string): Promise<UpdateResult> {
  const cached = readCache();
  if (cached) {
    return {
      current_version: currentVersion,
      latest_version: cached.latest_version,
      update_available: compareVersions(currentVersion, cached.latest_version) < 0,
      download_url: cached.download_url,
      release_notes: cached.release_notes,
      release_date: cached.release_date,
      error: null,
    };
  }

  try {
    const release = await fetchLatestRelease();
    const latest = release.tag_name.replace(/^v/i, "");
    writeCache({
      timestamp: nowSecs(),
      latest_version: latest,
      download_url: release.html_url ?? null,
      release_notes: release.body ?? null,
      release_date: release.published_at ?? null,
    });
    return {
      current_version: currentVersion,
      latest_version: latest,
      update_available: compareVersions(currentVersion, latest) < 0,
      download_url: release.html_url ?? null,
      release_notes: release.body ?? null,
      release_date: release.published_at ?? null,
      error: null,
    };
  } catch (err) {
    return {
      current_version: currentVersion,
      latest_version: currentVersion,
      update_available: false,
      download_url: null,
      release_notes: null,
      release_date: null,
      error: err instanceof Error ? err.message : String(err),
    };
  }
}

export function updatesCacheLabel(): string {
  return `${DISPLAY_NAME} update cache → ${appCacheDir()}`;
}

File: Sources/MetaBurn/Utilities/Updates.swift
import Foundation

struct UpdateResult {
    let currentVersion: String
    let latestVersion: String
    let updateAvailable: Bool
    let downloadURL: String?
    let releaseNotes: String?
    let releaseDate: String?
    let error: String?
}

final class Updates {
    static let cacheDuration: TimeInterval = 3600
    static let userAgent = "metaburn-update-checker/1.0"

    static func cacheURL() -> URL {
        Paths.ensureCacheDirectory()
        return Paths.cacheDirectory().appendingPathComponent("update_check.json")
    }

    static func checkForUpdates(currentVersion: String) async -> UpdateResult {
        if let cached = readCache(currentVersion: currentVersion) {
            return cached
        }

        let url = URL(string: "https://api.github.com/repos/\(Brand.githubOrg)/\(Brand.githubRepo)/releases/latest")!
        var request = URLRequest(url: url)
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/vnd.github.v3+json", forHTTPHeaderField: "Accept")

        let sessionDelegate = GithubPinningDelegate()
        let session = URLSession(configuration: .ephemeral, delegate: sessionDelegate, delegateQueue: nil)
        defer { session.finishTasksAndInvalidate() }

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode < 400 else {
                return UpdateResult(
                    currentVersion: currentVersion,
                    latestVersion: currentVersion,
                    updateAvailable: false,
                    downloadURL: nil,
                    releaseNotes: nil,
                    releaseDate: nil,
                    error: "GitHub returned HTTP \((response as? HTTPURLResponse)?.statusCode ?? 0)"
                )
            }
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
            let tag = (json["tag_name"] as? String ?? "").replacingOccurrences(of: "^v", with: "", options: .regularExpression)
            let htmlURL = json["html_url"] as? String
            let body = json["body"] as? String
            let publishedAt = json["published_at"] as? String

            writeCache(latestVersion: tag, downloadURL: htmlURL, releaseNotes: body, releaseDate: publishedAt)

            return UpdateResult(
                currentVersion: currentVersion,
                latestVersion: tag,
                updateAvailable: compareVersions(currentVersion, tag) < 0,
                downloadURL: htmlURL,
                releaseNotes: body,
                releaseDate: publishedAt,
                error: nil
            )
        } catch {
            return UpdateResult(
                currentVersion: currentVersion,
                latestVersion: currentVersion,
                updateAvailable: false,
                downloadURL: nil,
                releaseNotes: nil,
                releaseDate: nil,
                error: error.localizedDescription
            )
        }
    }

    static func readCache(currentVersion: String) -> UpdateResult? {
        let url = cacheURL()
        guard let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let timestamp = json["timestamp"] as? TimeInterval,
              Date().timeIntervalSince1970 - timestamp < cacheDuration,
              let latest = json["latest_version"] as? String else { return nil }

        let downloadURL = (json["download_url"] is NSNull) ? nil : (json["download_url"] as? String)
        let releaseNotes = (json["release_notes"] is NSNull) ? nil : (json["release_notes"] as? String)
        let releaseDate = (json["release_date"] is NSNull) ? nil : (json["release_date"] as? String)

        return UpdateResult(
            currentVersion: currentVersion,
            latestVersion: latest,
            updateAvailable: compareVersions(currentVersion, latest) < 0,
            downloadURL: downloadURL,
            releaseNotes: releaseNotes,
            releaseDate: releaseDate,
            error: nil
        )
    }

    static func writeCache(latestVersion: String, downloadURL: String?, releaseNotes: String?, releaseDate: String?) {
        let payload: [String: Any] = [
            "timestamp": Date().timeIntervalSince1970,
            "latest_version": latestVersion,
            "download_url": downloadURL ?? NSNull(),
            "release_notes": releaseNotes ?? NSNull(),
            "release_date": releaseDate ?? NSNull()
        ]
        if let data = try? JSONSerialization.data(withJSONObject: payload, options: .prettyPrinted) {
            try? data.write(to: cacheURL(), options: .atomic)
        }
    }

    static func compareVersions(_ a: String, _ b: String) -> Int {
        let parse: (String) -> [Int] = { version in
            version.replacingOccurrences(of: "^v", with: "", options: .regularExpression)
                .split(separator: ".")
                .compactMap { Int($0) }
        }
        let pa = parse(a), pb = parse(b)
        for i in 0..<max(pa.count, pb.count) {
            let x = i < pa.count ? pa[i] : 0
            let y = i < pb.count ? pb[i] : 0
            if x < y { return -1 }
            if x > y { return 1 }
        }
        return 0
    }
}
